src/codegen/mod.rs

```diff
@@ -563,131 +828,232 @@ fn gen_par_for(
         par_section
     };
 
-    cu::Stmt::Seq(Box::new(body), Box::new(sync_stmt))
+    cu::Stmt::Seq (
+        Box::new(body),
+        Box::new(sync_stmt),
+    )
+}
+
+fn build_seq(mut stmts: Vec<cu::Stmt>) -> cu::Stmt {
+    if stmts.len() <= 1 {
+        panic!("needs at least two stmts");
+    } else if stmts.len() == 2 {
+        cu::Stmt::Seq(
+            Box::new(stmts.remove(0)),
+            Box::new(stmts.remove(0)),
+        )
+    } else {
+        cu::Stmt::Seq(
+            Box::new(stmts.remove(0)),
+            Box::new(build_seq(stmts)),
+        )
+    }
+}
+
+fn gen_checked_stmt(
```

Comment:
We should talk about this function. It seems like a partial duplication of `gen_stmt`.